generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/Resolver.java

```diff
@@ -320,8 +321,11 @@ static String getSha1Url(String url, String extension) {
   private String downloadSha1(Rule rule) {
     String sha1Url = getSha1Url(rule.getUrl(), rule.getArtifact().getExtension());
     try {
-      HttpURLConnection connection = (HttpURLConnection) new URL(sha1Url).openConnection();
-      connection.setInstanceFollowRedirects(true);
+      URLConnection connection = new URL(sha1Url).openConnection();
+      if (connection instanceof HttpURLConnection) {
```

Comment:
Thank you again. 

In the situation that it is not an `HttpURLConnection`, it would be a [`JarURLConnection`](https://docs.oracle.com/javase/7/docs/api/java/net/JarURLConnection.html). I doubt this should occur, so can you throw an exception? 

Additionally, can you separate the sha functionality into a separate class, and add a simple test case for when this URLConnection fails? 

`Resolver.java` has gotten fairly bloated, and I am trying to steer it so that `Resolver.java` has less functionality backed in. (see `VersionResolver.java` as an example)

Reply:
The connection can be an instance of any type that the [`URLStreamHandler`](https://docs.oracle.com/javase/8/docs/api/index.html?java/net/URLStreamHandler.html) for the protocol returns. For a file, this appears to be `sun.net.www.protocol.file.FileURLConnection` (at least on the JRE I'm using).

Both the sha and pom files will have the protocol `file` if the repository has a `file` URL.